In [1]:
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

# Part 1

In [2]:
random_data = np.random.uniform(low=-5000, high=5000, size=80)
print("type of data:", type(random_data))
print("type of elem:", random_data.dtype)

type of data: <class 'numpy.ndarray'>
type of elem: float64


In [3]:
# this function round data for example convert 1.6 -> to 2 and -2.8 to 3
random_data_rounded = np.round(random_data)

* ### `float` and `uint` (unsigned integer) are not appropriate
* ### `int8` is for $[-128, 127]$
* ### `int16` is for $[-32768, 32767]$
* ### `int32` or `int64` reserve more than our need size in disk
## so i choose `int16`

In [4]:
random_data_rounded = random_data_rounded.astype(np.int16)
print("type of elem:", random_data_rounded.dtype)

type of elem: int16


In [5]:
# -5000 to 5000 has 10001 numbers so:
# x + 5000 -> / 10001.0 (for cast to float) -> * 256 -> convert to int16
converted_random_data = ((random_data_rounded + 5000) / 10001.0 * 256).astype(np.int16)
print("type of elem:", converted_random_data.dtype)

type of elem: int16


In [6]:
reshaped_random_data = converted_random_data.reshape(8, 10)
print(f"data:\n{reshaped_random_data}")

data:
[[224 208  39 244  94 132 168 219 208  37]
 [157 147 242 164 199  38 191 124  10 219]
 [199 145 103 116 232  66 242 128  15 132]
 [ 95 121 132  27 186  88 223 114 143 103]
 [ 34 107 111 117  11  27 250 224 196  58]
 [204 101   6 100 234  97  68 124  37  58]
 [240 228 225 175 105 214 150 247 105 249]
 [185   8  63 104 178 207 160 217 118 127]]


In [7]:
reshaped_random_data_int8 = reshaped_random_data.astype(np.int8)
print(f"int8 of data:\n{reshaped_random_data_int8}")

int8 of data:
[[ -32  -48   39  -12   94 -124  -88  -37  -48   37]
 [ -99 -109  -14  -92  -57   38  -65  124   10  -37]
 [ -57 -111  103  116  -24   66  -14 -128   15 -124]
 [  95  121 -124   27  -70   88  -33  114 -113  103]
 [  34  107  111  117   11   27   -6  -32  -60   58]
 [ -52  101    6  100  -22   97   68  124   37   58]
 [ -16  -28  -31  -81  105  -42 -106   -9  105   -7]
 [ -71    8   63  104  -78  -49  -96  -39  118  127]]


## in computer integer represent by [Two’s Complement method](https://en.wikipedia.org/wiki/Two%27s_complement)
## so after converting from `int16` to `int8`, `int8`  keeps the lowest 8 bits
* ### for example `224` is `0000000011100000` in `int16` after casting to `int8` keeps `11100000` that is `-32`